### ISSUE TITLE
`wrong-numbe-of-arguments` error from backported `find-sibling-file` command

### ISSUE THREAD

**Issue by manuel-arguelles** [doomemacs/doomemacs#7795]
### I confirm that...

- [X] I have searched the [issue tracker](https://github.com/doomemacs/doomemacs/issues), [documentation](https://docs.doomemacs.org), [FAQ](https://docs.doomemacs.org/-/faq), [Discourse](https://discourse.doomemacs.org), and [Google](https://google.com), in case this issue has already been reported/resolved.
- [X] I have read "[How to Debug Issues](https://doomemacs.org/d/how2debug)", and will use it to provide as much information about this issue as possible.
- [X] The issue can be reproduced on the **latest** available commit of Doom.
- [X] The issue can be reproduced on a stable release of Emacs, such as 27, 28, or 29. *(Unstable versions end in .50, .60, or .9x)*

### Expected behavior

When calling `SPC p o` the behavior of find-sibling-file should work similar to projectile-find-other-file. For c/c++ it should jump from the header file to the implementation and back.

### Current behavior

When calling find-sibling-file from a cpp implementation file (.cpp) the function fails with `Wrong number of arguments: (1 . 2), 3` the backtrace is:

```
Debugger entered--Lisp error: (wrong-number-of-arguments (1 . 2) 3)
  file-expand-wildcards("module.\\(h\\|hh\\|hpp\\)" nil t)
  (mapcar #'expand-file-name (file-expand-wildcards expansion nil t))
  (nconc results (mapcar #'expand-file-name (file-expand-wildcards expansion nil t)))
  (setq results (nconc results (mapcar #'expand-file-name (file-expand-wildcards expansion nil t))))
  (let ((default-directory (substring file 0 (car match-data)))) (setq results (nconc results (mapcar #'expand-file-name (file-expand-wildcards expansion nil t)))))
  (let ((expansion (car --dolist-tail--))) (let ((start 0)) (while (string-match "\\\\\\([&0-9]+\\)" expansion start) (let ((index (string-to-number (match-string 1 expansion)))) (setq start (match-end 0) expansion (replace-match (substring file (elt match-data ...) (elt match-data ...)) t t expansion))))) (let ((default-directory (substring file 0 (car match-data)))) (setq results (nconc results (mapcar #'expand-file-name (file-expand-wildcards expansion nil t))))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (while --dolist-tail-- (let ((expansion (car --dolist-tail--))) (let ((start 0)) (while (string-match "\\\\\\([&0-9]+\\)" expansion start) (let ((index (string-to-number ...))) (setq start (match-end 0) expansion (replace-match (substring file ... ...) t t expansion))))) (let ((default-directory (substring file 0 (car match-data)))) (setq results (nconc results (mapcar #'expand-file-name (file-expand-wildcards expansion nil t))))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  (let ((--dolist-tail-- expansions)) (while --dolist-tail-- (let ((expansion (car --dolist-tail--))) (let ((start 0)) (while (string-match "\\\\\\([&0-9]+\\)" expansion start) (let ((index ...)) (setq start (match-end 0) expansion (replace-match ... t t expansion))))) (let ((default-directory (substring file 0 (car match-data)))) (setq results (nconc results (mapcar #'expand-file-name (file-expand-wildcards expansion nil t))))) (setq --dolist-tail-- (cdr --dolist-tail--)))))
  (let ((match-data (match-data))) (let ((--dolist-tail-- expansions)) (while --dolist-tail-- (let ((expansion (car --dolist-tail--))) (let ((start 0)) (while (string-match "\\\\\\([&0-9]+\\)" expansion start) (let (...) (setq start ... expansion ...)))) (let ((default-directory (substring file 0 ...))) (setq results (nconc results (mapcar ... ...)))) (setq --dolist-tail-- (cdr --dolist-tail--))))))
  (progn (let ((match-data (match-data))) (let ((--dolist-tail-- expansions)) (while --dolist-tail-- (let ((expansion (car --dolist-tail--))) (let ((start 0)) (while (string-match "\\\\\\([&0-9]+\\)" expansion start) (let ... ...))) (let ((default-directory ...)) (setq results (nconc results ...))) (setq --dolist-tail-- (cdr --dolist-tail--)))))))
  (if (string-match match file) (progn (let ((match-data (match-data))) (let ((--dolist-tail-- expansions)) (while --dolist-tail-- (let ((expansion ...)) (let (...) (while ... ...)) (let (...) (setq results ...)) (setq --dolist-tail-- (cdr --dolist-tail--))))))))
  (let ((match x4) (expansions x5)) (if (string-match match file) (progn (let ((match-data (match-data))) (let ((--dolist-tail-- expansions)) (while --dolist-tail-- (let (...) (let ... ...) (let ... ...) (setq --dolist-tail-- ...))))))))
  (let* ((x4 (car-safe x3)) (x5 (cdr-safe x3))) (let ((match x4) (expansions x5)) (if (string-match match file) (progn (let ((match-data (match-data))) (let ((--dolist-tail-- expansions)) (while --dolist-tail-- (let ... ... ... ...))))))))
  (progn (ignore (consp x3)) (let* ((x4 (car-safe x3)) (x5 (cdr-safe x3))) (let ((match x4) (expansions x5)) (if (string-match match file) (progn (let ((match-data ...)) (let (...) (while --dolist-tail-- ...))))))))
  (let ((x3 (car --dolist-tail--))) (progn (ignore (consp x3)) (let* ((x4 (car-safe x3)) (x5 (cdr-safe x3))) (let ((match x4) (expansions x5)) (if (string-match match file) (progn (let (...) (let ... ...))))))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (while --dolist-tail-- (let ((x3 (car --dolist-tail--))) (progn (ignore (consp x3)) (let* ((x4 (car-safe x3)) (x5 (cdr-safe x3))) (let ((match x4) (expansions x5)) (if (string-match match file) (progn (let ... ...)))))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  (let ((--dolist-tail-- (or rules find-sibling-rules))) (while --dolist-tail-- (let ((x3 (car --dolist-tail--))) (progn (ignore (consp x3)) (let* ((x4 (car-safe x3)) (x5 (cdr-safe x3))) (let ((match x4) (expansions x5)) (if (string-match match file) (progn ...))))) (setq --dolist-tail-- (cdr --dolist-tail--)))))
  (let ((results nil)) (let ((--dolist-tail-- (or rules find-sibling-rules))) (while --dolist-tail-- (let ((x3 (car --dolist-tail--))) (progn (ignore (consp x3)) (let* ((x4 ...) (x5 ...)) (let (... ...) (if ... ...)))) (setq --dolist-tail-- (cdr --dolist-tail--))))) (delete file (delete-dups results)))
  find-sibling-file-search("/tmp/example/src/module.cpp" (("/\\([^/]+\\)\\.c\\(c\\|pp\\)\\'" "\\1.\\(h\\|hh\\|hpp\\)") ("/\\([^/]+\\)\\.\\(\\(s[ac]\\|le\\)ss\\|styl\\)\\'" "\\1\\.css")))
  (let ((siblings (find-sibling-file-search (expand-file-name file) find-sibling-rules))) (cond ((null siblings) (user-error "Couldn't find any sibling files")) ((length= siblings 1) (find-file (car siblings))) (t (let ((relatives (mapcar #'... siblings))) (find-file (completing-read (format-prompt "Find file" (car relatives)) relatives nil t nil nil (car relatives)))))))
  find-sibling-file("/tmp/example/src/module.cpp")
  funcall-interactively(find-sibling-file "/tmp/example/src/module.cpp")
  call-interactively(find-sibling-file nil nil)
  command-execute(find-sibling-file)
```

When called from a cpp header file (.h) the result is `Couldn't find any sibling file`.

This is in Emacs 28.1

### Steps to reproduce

1. Open Emacs (28.1)
2. Clone the example project that shows the case: https://github.com/manuel-arguelles/example
3. Import the project and try find-sibling-file

### System Information

https://termbin.com/51gx

**Comment by hlissner** [comment]
Would you mind testing this? 

```elisp
;;; add to $DOOMDIR/config.el

(defadvice! fixed--find-sibling-file-search (file &optional rules)
  :override #'find-sibling-file-search
  (let ((results nil))
    (pcase-dolist (`(,match . ,expansions) (or rules find-sibling-rules))
      (when (string-match match file)
        (let ((match-data (match-data)))
          (dolist (expansion expansions)
            (let ((start 0))
              (while (string-match "\\\\\\([&0-9]+\\)" expansion start)
                (let ((index (string-to-number (match-string 1 expansion))))
                  (setq start (match-end 0)
                        expansion
                        (replace-match
                         (substring file
                                    (elt match-data (* index 2))
                                    (elt match-data (1+ (* index 2))))
                         t t expansion)))))
            (let ((default-directory (substring file 0 (car match-data))))
              (setq results
                    (nconc
                     results
                     (mapcar #'expand-file-name
                             (letf! (defun wildcard-to-regexp (wildcard)
                                      (concat "\\`" wildcard "\\'"))
                               (file-expand-wildcards expansion nil))))))))))
    (delete file (delete-dups results))))
```

If it works, I'll have a fix up soon after.

**Comment by manuel-arguelles** [comment]
Thanks, it works from .cpp files to .h in the same directory, from the example code: from src/module.cpp it jumps to src/module.h, however it doesn't work from src/module.h to src/module.cpp (Couldn't find any sibling files), also from files in different directories (in both ways), from src/example.cpp fails to jumpt to include/example.h (Couldn't find any sibling files) and from include/example.h fails to find src/example.cpp

**Comment by manuel-arguelles** [comment]
Thanks @hlissner I can confirm it is working now, however only for files in the same directory (which is different from projectile-find-other-file, which was project wide). Thanks!

If projectile gets dropped, it would be nice to have similar function (`better-jumper-jump-backward` works fine but but I see it more like a history back than a method to jump to a sibling file. Anyways, it's not a big deal. I'd try to upgrade emacs to see how the native implementation behaves.

**Comment by hlissner** [comment]
I haven't pushed a fix for either issue yet. I'll close this once I have.

>  however only for files in the same directory (which is different from projectile-find-other-file, which was project wide)

That's true. I do want parity between these two so I'll address this at some point.

**Comment by hlissner** [comment]
The two original issues should be fixed in 38a3adc and 19e2776. The recursive search, though, will likely need a separate command or smarter processing of `find-sibling-rules`. I'll look into that after we've fully switched over to project.el (I've added it to doomemacs/core#1). Thanks for bringing this to my attention in any case!
